### DISCUSSION TITLE
Programming STM32L4 over USB DFU

### DISCUSSION THREAD

**Discussion by MikeGHAtCOM** [K-Society/KSociety.SharpCubeProgrammer#136] [Q&A]
I've got a project based on the quickstart, and this line causes me issues

`connectionResult = programmer.DownloadFile(tfn, "0x08004800", 0, 1);`

The output says

`Opening and parsing file: tmp2asvyu.hex: Message: {0}
  File          : tmp2e1xpo.hex: Message: {0}
  Size          : 165.53 KB : Message: {0}
  Address       : 0x08004800 : Message: {0}
Erasing Segment <0> Address <0x08004800> Size <148268>Bytes: Message: {0}
Erasing memory corresponding to segment 0:: Message: {0}
Error: failed to erase memory: Message: {0}
Disconnected from device.: Message: {0}`

Using the same hex file in the STM32CubeProgrammer with the same device programs OK, so I think the hex file is ok

`[...]15:29:48 : Time elapsed during the read operation is: 00:00:00.056
  15:29:49 : Memory Programming ...
  15:29:49 : Opening and parsing file: tmp2asvyu.hex
  15:29:49 :   File          : tmp2asvyu.hex
  15:29:49 :   Size          : 165.53 KB 
  15:29:49 :   Address       : 0x08004800 
  15:29:49 : Erasing memory corresponding to segment 0:
  15:29:49 : erasing sector 0009 @: 0x08004800 done
  15:29:49 : erasing sector 0010 @: 0x08005000 done
  15:29:49 : erasing sector 0011 @: 0x08005800 done
  15:29:49 : erasing sector 0012 @: 0x08006000 done
  15:29:49 : erasing sector 0013 @: 0x08006800 done
  15:29:49 : erasing sector 0014 @: 0x08007000 done
  15:29:49 : erasing sector 0015 @: 0x08007800 done
[...]`

**Comment by maniglia** [comment]
Hello @MikeGHAtCOM 
Could you try using a bin file instead of the hex, please?
I don't have time at the moment but I'll do some tests later.
I'm using DFU mode in production with bin files, at least with bin files it should work.

**Comment by MikeGHAtCOM** [comment]
using the bin file also gives the same error

USB speed   : Full Speed (12MBit/s): Message: {0}
Manuf. ID   : STMicroelectronics: Message: {0}
Product ID  : STM32 DownLoad Firmware Update: Message: {0}
SN          : 206936745830: Message: {0}
DFU protocol: 1.1: Message: {0}
Board       : --: Message: {0}
IAP: Message: {0}
Memory Programming ...: Message: {0}
Opening and parsing file: tmp2lvkdz.bin: Message: {0}
  File          : tmp2lvkdz.bin: Message: {0}
  Size          : 144.79 KB : Message: {0}
  Address       : 0x08004800 : Message: {0}
Erasing Segment <0> Address <0x08004800> Size <148268>Bytes: Message: {0}
Erasing memory corresponding to segment 0:: Message: {0}
Error: failed to erase memory: Message: {0}

STM32CubeProgrammer on the same one says:

10:00:26 : Opening and parsing file: tmp2lvkdz.bin
  10:00:26 :   File          : tmp2lvkdz.bin
  10:00:26 :   Size          : 144.79 KB
  10:00:26 :   Address       : 0x08004800 
  10:00:26 : Erasing memory corresponding to segment 0:
  10:00:26 : erasing sector 0009 @: 0x08004800 done
  10:00:26 : erasing sector 0010 @: 0x08005000 done
  10:00:26 : erasing sector 0011 @: 0x08005800 done
  10:00:26 : erasing sector 0012 @: 0x08006000 done
  10:00:26 : erasing sector 0013 @: 0x08006800 done

My STM32CubeProgrammer is 2.13.0

**Comment by MikeGHAtCOM** [comment]
to establish it wasn't running an old version of CubeProgrammer, I've reproduced it with 

STM32CubeProgrammer API v2.16.0 | Windows-64Bits 
and
STM32CubeProgrammer API v2.16.0 | Windows-32Bits

**Comment by maniglia** [comment]
I think there is a bug in the latest version of CubeProgrammer...
Try using the _MassErase_ function before the _DownloadFile_ function.
